### PR TITLE
Add Air Quality Cluster Definition

### LINKED ISSUE #26194: Add Air Quality Cluster Definition
# 📖 Description
To add full support for the Air Quality Cluster to the SDK a cluster definition needs to be added. 

## 🔎 Details
A cluster definition is an xml file which describes the cluster according to the spec.

## ✅ Tasks 
- [x] Create `<cluster-name>-cluster.xml` under `src/app/zap-templates/zcl/data-model/chip`
- [x] Add file reference to `.github/workflows/tests.yaml`
- [x] Add file reference to `scripts/rules.matterlint`
- [x] Add file reference to `src/app/zap-templates/zcl/data-model/all.xml`
- [x] Add file reference to `src/app/zap-templates/zcl/zcl-with-test-extensions.json`
- [x] Add file reference to `src/app/zap-templates/zcl/zcl.json`
- [x] Enable new cluster to the Darwin and Android clients (`controller-clusters.zap`)
- [x] Add all new clusters to `src/app/zap_cluster_list.json`
- [x] Run `./scripts/tools/zap_regen_all.py`

## 🔗 Links/References
- [Wiki from Petru on how to add a new device type](https://groups.csa-iot.org/wg/matter-tsg/wiki/howto-add-a-new-device-type)

### PR DESCRIPTION
**Pull request by p0fi**
# 📖 Description
This PR adds the air quality cluster definition (`air-quality-cluster.xml`) zap template. 

## 🔎 Details
- Template is the specification of the open PR (https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/6559)
- Template is added to various lists in various files (see #26194)
- Code generation ran with `zap_regen_all.py`

## 🧪 How Has This Been Tested
- chip-tool was built using `./scripts/examples/gn_build_example.sh examples/chip-tool out/chip-tool`

## 🔗 Links/References
- closes #26194

### CODE REVIEW

File: src/controller/data_model/controller-clusters.zap [hunk 1]
```diff
@@ -16861,7 +17025,7 @@
               "side": "server",
               "type": "LineupInfoStruct",
               "included": 1,
-              "storageOption": "RAM",
+              "storageOption": "External",
```
Comment (andy31415) [inline src/controller/data_model/controller-clusters.zap:17028]: these changes from RAM to External seem unexpected. At the same time I imagine controller-clusters does not care about them. Would feel safer if these deltas would not exist though.

File: src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    <field name="Extremely poor" mask="0x08" />
+  </bitmap>
+
+  <!-- Cluster Data Types -->
+  <enum name="AirQualityEnum" type="ENUM8">
+    <cluster code="0x005B" />
+    <item value="0" name="Unknown" />
+    <item value="1" name="Good" />
+    <item value="2" name="Fair" />
+    <item value="3" name="Poor" />
```
Comment (andy31415) [inline src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml:47]: Moderate seems missing (at least I have it in my latest spec copy)

File: src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    <name>Air Quality</name>
+    <domain>Measurement &amp; Sensing</domain>
+    <description>Attributes for reporting air quality classification</description>
+    <code>0x005B</code>
+    <define>AIR_QUALITY_CLUSTER</define>
+    <client tick="false" init="false">true</client>
+    <server tick="false" init="false">true</server>
+
+    <!-- Attributes -->
+    <attribute side="server" code="0x0000" define="AIR_QUALITY" type="AirQualityEnum" min="0" max="5" writable="false" isNullable="false" default="0" optional="false">AirQuality</attribute>
```
Comment (andy31415) [inline src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml:29]: How about `SupportedLevels` ?

File: src/controller/data_model/controller-clusters.zap (hunk 1 above)
Comment (bzbarsky-apple) [reply src/controller/data_model/controller-clusters.zap:17028]: These changes are expected if there have been changes to some metadata about which things are forced to be AttributeAccessInterface and then no one ever updated this file (by running ZAP UI or running the update-zap-file script).

File: src/app/zap-templates/zcl/data-model/all.xml [hunk 4]
```diff
@@ -75,4 +75,5 @@
    <xi:include href="draft/electrical-measurement-cluster.xml" />
    <xi:include href="draft/input-output-value-clusters.xml" />
    <xi:include href="draft/onoff-switch-configuration-cluster.xml" />
+   <xi:include href="chip/air-quality-cluster.xml" />
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/all.xml:78]: Please add this in alphabetical order like all the other bits here?

File: src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml (hunk 3 above)
Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml:29]: I assume this is based on https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/6559; it does say that above.

That unfortunately means you have to use a spec from the PR to compare to.

Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml:29]: I would expect the max here to be 6, which is "Extremely poor".

File: src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  </cluster>
+
+  <!-- Feature Map -->
+  <bitmap name="AirQualityFeature" type="BITMAP32">
+    <cluster code="0x005B" />
+    <field name="Fair" mask="0x01" />
+    <field name="Moderate" mask="0x02" />
+    <field name="Very poor" mask="0x04" />
+    <field name="Extremely poor" mask="0x08" />
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml:38]: This matches the spec as it is in the spec PR, but note the comments there.

File: src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml:47]: Yes, that likely explains why this ended up at 5, not 6, for the max.

File: src/app/zap-templates/zcl/zcl-with-test-extensions.json [hunk 6]
```diff
@@ -91,7 +91,8 @@
         "types/color-control.xml",
         "types/door-lock.xml",
         "types/occupancy-sensing.xml",
-        "types/thermostat-user-interface-configuration.xml"
+        "types/thermostat-user-interface-configuration.xml",
+        "air-quality-cluster.xml"
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/zcl-with-test-extensions.json:95]: Again, please add in alphabetical order above?

File: src/app/zap-templates/zcl/zcl.json [hunk 7]
```diff
@@ -89,7 +89,8 @@
         "types/color-control.xml",
         "types/door-lock.xml",
         "types/occupancy-sensing.xml",
-        "types/thermostat-user-interface-configuration.xml"
+        "types/thermostat-user-interface-configuration.xml",
+        "air-quality-cluster.xml"
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/zcl.json:93]: Alphabetical order.

File: src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml (hunk 2 above)
Comment (p0fi) [reply src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml:47]: yep, sorry gonna fix this! 👍

File: src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml (hunk 3 above)
Comment (p0fi) [reply src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml:29]: I would also loved to have this PR merged but it seems policy that during the ballot nothing gets merged 🙁

Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml:29]: Right, that was just context for @andy31415 so he knows what he should look for....

File: src/controller/data_model/controller-clusters.zap (hunk 1 above)
Comment (p0fi) [reply src/controller/data_model/controller-clusters.zap:17028]: So I assume this change is okay then and the discussion could be marked as resolved @bzbarsky-apple?

File: src/app/zap-templates/zcl/zcl-with-test-extensions.json [hunk 8]
```diff
@@ -44,15 +45,16 @@
...
         "group-key-mgmt-cluster.xml",
+        "groups-cluster.xml",
         "identify-cluster.xml",
         "illuminance-measurement-cluster.xml",
         "input-output-value-clusters.xml",
         "keypad-input-cluster.xml",
         "level-control-cluster.xml",
         "localization-configuration-cluster.xml",
         "low-power-cluster.xml",
+        "matter-devices.xml",
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/zcl-with-test-extensions.json:57]: Again, this one, not being a cluster XML, should perhaps go at the end.

File: src/app/zap-templates/zcl/zcl-with-test-extensions.json [hunk 9]
```diff
@@ -82,16 +84,15 @@
...
-        "unit-localization-cluster.xml",
-        "wake-on-lan-cluster.xml",
-        "wifi-network-diagnostics-cluster.xml",
-        "window-covering.xml",
-        "matter-devices.xml",
         "types/color-control.xml",
         "types/door-lock.xml",
         "types/occupancy-sensing.xml",
-        "types/thermostat-user-interface-configuration.xml"
+        "types/thermostat-user-interface-configuration.xml",
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/zcl-with-test-extensions.json:90]: And here, what I would say we should do is, in order:

1. Clusters in alphabetical order
2. matter-devices
3. types/* in alphabetical order

File: src/app/zap-templates/zcl/zcl.json [hunk 10]
```diff
@@ -43,15 +44,16 @@
...
         "group-key-mgmt-cluster.xml",
+        "groups-cluster.xml",
         "identify-cluster.xml",
         "illuminance-measurement-cluster.xml",
         "input-output-value-clusters.xml",
         "keypad-input-cluster.xml",
         "level-control-cluster.xml",
         "localization-configuration-cluster.xml",
         "low-power-cluster.xml",
+        "matter-devices.xml",
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/zcl.json:56]: Same comments as for zcl-test-with-extensions.json

File: src/app/zap-templates/zcl/zcl-with-test-extensions.json (hunk 9 above)
Comment (p0fi) [reply src/app/zap-templates/zcl/zcl-with-test-extensions.json]: Reverted the sorting part of the PR

File: src/app/zap-templates/zcl/zcl.json (hunk 10 above)
Comment (p0fi) [reply src/app/zap-templates/zcl/zcl.json]: Reverted the sorting part of the PR

File: src/app/zap-templates/zcl/zcl-with-test-extensions.json (hunk 8 above)
Comment (p0fi) [reply src/app/zap-templates/zcl/zcl-with-test-extensions.json]: Reverted the sorting part of the PR

File: src/app/zap-templates/zcl/data-model/all.xml [hunk 11]
```diff
@@ -22,31 +24,32 @@
...
    <xi:include href="chip/group-key-mgmt-cluster.xml" />
+   <xi:include href="chip/groups-cluster.xml" />
    <xi:include href="chip/identify-cluster.xml" />
    <xi:include href="chip/illuminance-measurement-cluster.xml" />
-   <xi:include href="chip/level-control-cluster.xml" />
    <xi:include href="chip/keypad-input-cluster.xml" />
+   <xi:include href="chip/level-control-cluster.xml" />
    <xi:include href="chip/localization-configuration-cluster.xml" />
    <xi:include href="chip/low-power-cluster.xml" />
+   <xi:include href="chip/matter-devices.xml" />
```
[the remark this answers is not in the document]
Comment (p0fi) [reply src/app/zap-templates/zcl/data-model/all.xml]: Reverted the sorting part of the PR

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    <define>AIR_QUALITY_CLUSTER</define>
+    <client tick="false" init="false">true</client>
+    <server tick="false" init="false">true</server>
+
+    <!-- Attributes -->
+    <attribute side="server" code="0x0000" define="AIR_QUALITY" type="AirQualityEnum" min="0" max="6" writable="false" isNullable="false" default="0" optional="false">AirQuality</attribute>
+  </cluster>
+
+  <!-- Feature Map -->
+  <bitmap name="Features" type="BITMAP32">
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/air-quality-cluster.xml:33]:
```suggestion
  <bitmap name="Feature" type="BITMAP32">
```
Sorry for the back-and-forth here, but this probably reads better in terms of the resulting APIs.

**Comment by andy31415** [comment]
@p0fi this needs merge with master.

**Review by andy31415 — APPROVED** [review]

**Pull request merged**